runtime/ops/command.rs

```diff
@@ -0,0 +1,324 @@
+  command.args(args.args);
+
+  if let Some(cwd) = args.cwd {
+    command.current_dir(cwd);
+  }
+
+  if args.clear_env {
+    command.env_clear();
+  }
+  command.envs(args.env);
+
+  #[cfg(unix)]
+  if let Some(gid) = args.gid {
+    super::check_unstable(state, "Deno.Command.gid");
+    command.gid(gid);
+  }
+  #[cfg(unix)]
+  if let Some(uid) = args.uid {
+    super::check_unstable(state, "Deno.Command.uid");
+    command.uid(uid);
+  }
+  #[cfg(unix)]
+  unsafe {
+    command.pre_exec(|| {
+      libc::setgroups(0, std::ptr::null());
+      Ok(())
+    });
+  }
+
+  command.stdin(subprocess_stdio_map(&args.stdio.stdin)?);
+  command.stdout(subprocess_stdio_map(&args.stdio.stdout)?);
+  command.stderr(subprocess_stdio_map(&args.stdio.stderr)?);
+
+  Ok(command)
+}
+
+#[derive(Serialize)]
+#[serde(rename_all = "camelCase")]
+struct Child {
+  rid: ResourceId,
+  pid: u32,
+  stdin_rid: Option<ResourceId>,
+  stdout_rid: Option<ResourceId>,
+  stderr_rid: Option<ResourceId>,
+}
+
+#[op]
+fn op_command_spawn(
+  state: &mut OpState,
+  args: CommandArgs,
+) -> Result<Child, AnyError> {
+  let mut command = tokio::process::Command::from(create_command(state, args)?);
+  // TODO(@crowlkats): allow detaching processes.
+  //  currently deno will orphan a process when exiting with an error or Deno.exit()
+  // We want to kill child when it's closed
+  command.kill_on_drop(true);
+
+  let mut child = command.spawn()?;
+  let pid = child.id().expect("Process ID should be set.");
+
+  let stdin_rid = child
+    .stdin
+    .take()
+    .map(|stdin| state.resource_table.add(ChildStdinResource::from(stdin)));
+
+  let stdout_rid = child
+    .stdout
+    .take()
+    .map(|stdout| state.resource_table.add(ChildStdoutResource::from(stdout)));
+
+  let stderr_rid = child
+    .stderr
+    .take()
+    .map(|stderr| state.resource_table.add(ChildStderrResource::from(stderr)));
+
+  let child_rid = state
+    .resource_table
+    .add(ChildResource(AsyncRefCell::new(child)));
+
+  Ok(Child {
+    rid: child_rid,
+    pid,
+    stdin_rid,
+    stdout_rid,
+    stderr_rid,
+  })
+}
+
+#[op]
+fn op_command_status(
+  state: &mut OpState,
+  rid: ResourceId,
+) -> Result<Option<CommandStatus>, AnyError> {
+  let resource = state.resource_table.get::<ChildResource>(rid)?;
+  let mut child = RcRef::map(resource, |r| &r.0).try_borrow_mut().unwrap();
+  Ok(child.try_wait()?.map(|status| status.into()))
+}
+
+#[op]
+async fn op_command_wait(
+  state: Rc<RefCell<OpState>>,
+  rid: ResourceId,
+  stdin_rid: Option<ResourceId>,
+) -> Result<CommandStatus, AnyError> {
+  let resource = state
+    .borrow_mut()
+    .resource_table
+    .take::<ChildResource>(rid)?;
+  let mut child = RcRef::map(resource, |r| &r.0).borrow_mut().await;
+  if let Some(stdin_rid) = stdin_rid {
+    let stdin = state
+      .borrow_mut()
+      .resource_table
+      .take::<ChildStdinResource>(stdin_rid)?;
+    child.stdin = Some(Rc::try_unwrap(stdin).unwrap().into_inner());
+  }
```

Comment:
I thought we had previously established that we do not want the auto-closing stdin behavior for the low level API?

Reply:
not that I am aware. also, once again:
> The stdin handle to the child process, if any, will be closed before waiting. This helps avoid deadlock: it ensures that the child does not block waiting for input from the parent, while the parent waits for the child to exit.

and since we want this similar to the rust API, we should do this imo.

Reply:
You might have not been part of this discussion, but we definitely established this. I can elaborate on why if you want.

Reply:
sure

Reply:
Essentially you want to allow users to perform the following:

```js
const child = Deno.spawnChild("foo", { stdin: "piped" });
Deno.stdin.readable.pipeTo(child.stdin).catch(e => console.error("piping failed", e))
const status = await child.status;
console.log("the child exited", status);
```

If we **always** auto-close stdin, this doesn't work anymore.

Also auto-closing behaviour is not what Node, or most other runtimes do. Tokio also only does it if you don't manually take control over the stdin handle (which we do, but then for some reason return it control before calling `Child#wait()`).

Reply:
Also this behaviour is totally inconsistent with `.output()`, which doesn't yield tokio control over the stdin handle, so it doesn't auto-close stdin when calling `Child#output()`.